library/src/image.cxx

```diff
@@ -351,6 +353,149 @@ std::vector<unsigned char> image::saveBuffer(SaveFormat format) const
   }
 }
 
+//----------------------------------------------------------------------------
+const f3d::image& image::toTerminalText(std::ostream& stream) const
+{
+  const int depth = this->getChannelCount();
+  if (this->getChannelType() != ChannelType::BYTE || depth < 3 || depth > 4)
+  {
+    throw std::invalid_argument("image must be byte RGB or RGBA");
+  }
+
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  const int width = dims[0];
+  const int height = dims[1];
+  const unsigned char* content = static_cast<unsigned char*>(this->getContent());
+
+  constexpr unsigned char alphaCutoff = 127;
+
+  /* Function to retrieve pixels so we can return:
+    - transparent black values for out-of-bounds coords,
+    - opaque alpha value when the image has no alpha channel.
+    Rendering with half blocks means 1 line of text represents 2 rows of pixels
+    so we _will_ attempt to access a line past the bottom if the height is not even.
+  */
+  const auto getPixel = [=](int x, int y)
+  {
+    if (x >= 0 && x < width && y >= 0 && y < height)
+    {
+      const size_t i = depth * ((height - 1 - y) * width + x);
+      const int rgb = content[i + 0] << 16 | content[i + 1] << 8 | content[i + 2];
```

Comment:
Why are you packing the values in an `int` instead of a `char[3]`?  
The bit manipulation looks useless to me, unless I'm missing something.

Reply:
we need to compare the RGB values to the previous/current background and foreground (that are either RGB or "not set") to optimize the escape sequences so it's easier to compare `int`s that can be unset to `-1` than to drag 3 `uchar`s + 1 `bool` around.
I guess the _right_ way to do it would be to go full `optional<triple<uchar, uchar, uchar>>` but that seems a bit heavy :/
What's your take?

Reply:
Aaaaah got it.